lib/media/track/trackpublication.js

```diff
@@ -0,0 +1,83 @@
+'use strict';
+
+const { EventEmitter } = require('events');
+const { buildLogLevels } = require('../../util');
+const { DEFAULT_LOG_LEVEL } = require('../../util/constants');
+const Log = require('../../util/log');
+let nInstances = 0;
+
+/**
+ * A {@link TrackPublication} represents a {@link Track} that
+ * has been published to a {@link Room}.
+ * @property {boolean} isTrackEnabled - whether the published {@link Track} is enabled
+ * @property {string} trackName - the {@link Track}'s name
+ * @property {Track.SID} trackSid - SID assigned to the published {@link Track}
+ * @emits TrackPublication#trackDisabled
+ * @emits TrackPublication#trackEnabled
+ */
+class TrackPublication extends EventEmitter {
+  constructor(isTrackEnabled, trackName, options) {
+    super();
+
+    options = Object.assign({
+      logLevel: DEFAULT_LOG_LEVEL
+    }, options);
+
+    const logLevels = buildLogLevels(options.logLevel);
+
+    Object.defineProperties(this, {
+      _instanceId: {
+        value: nInstances++
+      },
+      _isTrackEnabled: {
+        value: isTrackEnabled,
+        writable: true
+      },
+      _log: {
+        value: options.log || new Log('default', this, logLevels)
+      },
+      _trackSid: {
+        value: null,
+        writable: true
+      },
+      trackName: {
+        value: trackName
+      }
+    });
+  }
+
+  toString() {
+    return `[TrackPublication #${this._instanceId}: ${this.trackName}]`;
+  }
+
+  get isTrackEnabled() {
+    return this._isTrackEnabled;
+  }
+
+  set isTrackEnabled(isTrackEnabled) {
+    if (this._isTrackEnabled !== isTrackEnabled) {
+      this._isTrackEnabled = isTrackEnabled;
+      this.emit(this._isTrackEnabled ? 'trackEnabled' : 'trackDisabled');
+    }
+  }
```

Comment:
Can we avoid the side-effecting setter? Or at least move the side-effects to a setter for `_isTrackEnabled` instead of `isTrackEnabled`? IMO, `isTrackEnabled` should not have a setter.